templates/install_puppet.bat.erb

```diff
@@ -12,11 +12,22 @@ for /f "delims=" %%i in ('puppet config print --section agent environment') do s
 if exist %pid_path% del %pid_path%
 @echo %pid%> %pid_path%
 
+SET /A pid_checks_performed=0
+
 :wait_for_pid
 timeout /t 5 /nobreak > NUL
-FOR /F "tokens=*" %%A IN ('tasklist /FI "PID eq %AGENT_PID%" /NH') DO set _task=%%A
-echo %_task% | findstr "No tasks are running" >nul
-IF NOT %errorlevel% == 0 ( GOTO wait_for_pid )
+wmic path Win32_Process where handle=%AGENT_PID% get handle /format:textvaluelist | findstr /I "Handle=%AGENT_PID%"
+
+SET /A pid_checks_performed+=1
+REM Wait for a max of 120 seconds (or 24 iterations) before aborting the upgrade
+IF %errorlevel% == 0 IF %pid_checks_performed% == 24 (
+  REM Adding CustomSource allows EventCreate to work properly
+  reg add HKLM\SYSTEM\CurrentControlSet\services\eventlog\Application\Puppet /v CustomSource /t REG_DWORD /d 1 /f
+  EVENTCREATE /T ERROR /L APPLICATION /SO Puppet /ID 101 /D "Puppet agent upgrade failed while waiting for Puppet process [ID: %AGENT_PID%] to exit."
```

Comment:
That's a good question - thanks for pointing it out.

We only have 3 different error messages (which could theoretically be localized) defined in the Puppet event log.  

They were introduced back in this commit - https://github.com/puppetlabs/puppet/commit/d88f3e7c387eff270101604f2fec2e086f88ebdd

I intentionally selected a value that didn't conflict with any of these (really, the error codes should be unique to a particular failure type), with the hopes that we would later define 101 to be an upgrade failure.